### ISSUE TITLE
[BUG] SHAP values using treeExplainer and lightgbm for classification are wrong because of the different output type

### ISSUE THREAD

**Issue by ThomasBury** [Ekeany/Boruta-Shap#13]
Describe the bug
================
Hi, here a short description

 * TreeExplainer behaves differently given the passed model (namely, the output is not always of the same type, np.array or a list of np.array's). See [the lightgbm issue](https://github.com/slundberg/shap/issues/526)
 * SHAP and permutation importance should be computed on unseen data
 * SHAP importances are mean( | shap.values | ), so for classification, before taking the mean/sum, the abs value should be applied. 

To Reproduce
============
See from line 589 to 608

Expected behavior
=================
Something like the following worked for me (related to the PR I made for Boruta_Py). It can be "beautify".

```python
self.is_cat = 'catboost' in str(type(self.estimator))
self.is_lgb = 'lightgbm' in str(type(self.estimator))
```

```python
    def _is_tree_based(self):
        """
        checking if the estimator is tree-based (kernel SAP is too slow to be used here, unless using sampling)
        :return:
        """
        tree_based_models = ['lightgbm', 'xgboost', 'catboost', '_forest']
        condition = any(i in str(type(self.estimator)) for i in tree_based_models)
        return condition
```

```python

        if not self.classification:
             X_tr, X_tt, y_tr, y_tt = train_test_split(X, y, random_state=42)
        else:
             X_tr, X_tt, y_tr, y_tt = train_test_split(X, y, stratify=y, random_state=42)

        X_tr = pd.DataFrame(X_tr)
        X_tt = pd.DataFrame(X_tt)
        obj_feat = list(set(list(X_tr.columns)) - set(list(X_tr.select_dtypes(include=[np.number]))))

        if obj_feat:
            X_tr[obj_feat] = X_tr[obj_feat].astype('str').astype('category')
            X_tt[obj_feat] = X_tt[obj_feat].astype('str').astype('category')

        if self._is_tree_based():
            try:
                if self.is_cat:
                    model = self.estimator.fit(X_tr, y_tr, cat_features=obj_feat)
                else:
                    model = self.estimator.fit(X_tr, y_tr)

            except Exception as e:
                raise ValueError('Please check your X and y variable. The provided '
                                 'estimator cannot be fitted to your data.\n' + str(e))
            # build the explainer
            explainer = shap.TreeExplainer(model, feature_perturbation="tree_path_dependent")
            shap_values = explainer.shap_values(X_tt)
            # flatten to 2D if classification and lightgbm
            if is_classifier(self.estimator):
                if isinstance(shap_values, list):
                    # for lightgbm clf sklearn api, shap returns list of arrays
                    # https://github.com/slundberg/shap/issues/526
                    class_inds = range(len(shap_values))
                    shap_imp = np.zeros(shap_values[0].shape[1])
                    for i, ind in enumerate(class_inds):
                        shap_imp += np.abs(shap_values[ind]).mean(0)
                    shap_imp /= len(shap_values)
                else:
                    shap_imp = np.abs(shap_values).mean(0)
            else:
                shap_imp = np.abs(shap_values).mean(0)
        else:
            raise ValueError('Not a tree based model')
```

Screenshots
===========

Additional context
==================

**Comment by Ekeany** [comment]
Hi Thomas,

Sorry for the late reply I have been quite busy recently.

1. TreeExplainer behaves differently given the passed model (namely, the output is not always of the same type, np.array or a list of np.array's). See the lightgbm issue

Thanks for the heads up on this, I had noticed that SHAP returns weird types for different models but I missed this one. I will try and implement your solution as soon as possible.

2. SHAP and permutation importance should be computed on unseen data

I know this is a debated topic and I don't believe any research has been done. You obviously feel that we should explain the features with regards to how much they contribute to the models performance. Relative to the current scenario where we are just measure which features the model relies on most ?. 

3. SHAP importances are mean( | shap.values | ), so for classification, before taking the mean/sum, the abs value should be applied.

I believe I am currently doing this in the code but I maybe mistaken. The fact that negative numbers appear at the moment is due to the fact that I am standardizing the outputs.

**Comment by ThomasBury** [comment]
Hi Ekeany,

it's alright, we all are busy and doing this in our spare time. Hoping that Boruta_py will merge the Boruta_shap parts it currently lacks. It would be nice to have all in one and under the "scikit-contrib" banner, let's see. Thank you for your reply and your tidy package (much cleaner than my Boruta_py PR, however, if you find some useful material in it, do not hesitate to use it for Boruta_shap).


1. Indeed it's a bit misleading the way that SHAP returns either a np.array or a list. You can double-check my work-around, use it as is or "beautify" (it's kinda hacky). As you prefer 
Note that the lightgbm implementation is different from the SHAP one ( [see here for instance](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRegressor.html?highlight=predict#lightgbm.LGBMRegressor.predict))

2. You're right, it's debated ( [in the excellent IML, 5.5.2](https://christophm.github.io/interpretable-ml-book/feature-importance.html#feature-importance-data) pro and cons for train vs test). Here some thoughts: if you use train data, then you are more interested in the "mechanics" of the model, how the rules were derived and what is the impact of a particular predictor on this set of rules. If you use test data, then you are interested in the question "how does it generalize to unseen data" or how can I explain the predicted values to my management/Customer/Regulator (explain the score to a new patient, of a new financial instrument to the Regulators, etc.). There is [a great discussion here](https://compstat-lmu.github.io/iml_methods_limitations/pfi-data.html#introduction-to-test-vs.training-data). Bottom line both make sense, I was focused on the interpretability for regulation purposes and being consistent with the SHAP packages, which Authors recommend using test set. Could it be an optional argument (something like "in_sample =True") and then an if loop with and without train_test_split? Although, I doubt that there will be significant differences if the model does not overfit (so perhaps a little return on investment).

3. Correct, I misread, sorry about that.

**Comment by Ekeany** [comment]
Hi Thomas,

Yes it would be nice to see both packages consolidated under one name. However the owner of the scikit-learn-contrib seems very hesitant to add any extra dependencies or changes to the base functionality. Haha I am not sure about tidy code probably needs to be refactored if I am honest, software engineering isn't my strong suit.

With respect to the train vs test argument, those are some great resources you have provided. I definitely agree with your interpretation and use case for calculating the model importance on unseen data. Therefore I have added an optional argument train_or_test to the fit function in my most recent commit. If the argument is set to 'train' then the algorithm will act as normal computing the feature importance on already seen data. Whereas, if the argument is set to 'test' the feature importance is only calculated on the unseen data.

**Comment by ThomasBury** [comment]
Hi Ekeany,

yep, he doesn't like dependency, even not matplotlib (which I don't really get since Boruta relies on scikit-learn which is built on NumPy, SciPy, and matplotlib).

Anyway, I had a look and I think the SHAP format for classification and the train_or test issues are solved.
Therefore, I'm closing the issue. 

Thank you for the modifications you made.
